diff --git a/drivers/soc/mediatek/mtk-pmic-wrap.c b/drivers/soc/mediatek/mtk-pmic-wrap.c
index c725315..0bbdd10 100644
--- a/drivers/soc/mediatek/mtk-pmic-wrap.c
+++ b/drivers/soc/mediatek/mtk-pmic-wrap.c
@@ -1891,6 +1891,18 @@ static const struct of_device_id of_pwrap_match_tbl[] = {
 };
 MODULE_DEVICE_TABLE(of, of_pwrap_match_tbl);
 
+struct regmap *pwrap_node_to_regmap(struct device_node *np)
+{
+	struct platform_device *pdev;
+	struct pmic_wrapper *wrp;
+	pdev = of_find_device_by_node(np);
+	if (!pdev)
+		return ERR_PTR(-ENODEV);
+	wrp = platform_get_drvdata(pdev);
+	return wrp->regmap;
+}
+EXPORT_SYMBOL_GPL(pwrap_node_to_regmap);
+
 static int pwrap_probe(struct platform_device *pdev)
 {
 	int ret, irq;
diff --git a/drivers/soc/mediatek/mtk-scpsys.c b/drivers/soc/mediatek/mtk-scpsys.c
index 503222d..d104677 100644
--- a/drivers/soc/mediatek/mtk-scpsys.c
+++ b/drivers/soc/mediatek/mtk-scpsys.c
@@ -237,7 +237,8 @@ static int scpsys_power_on(struct generic_pm_domain *genpd)
 		 * MT7622_POWER_DOMAIN_WB and thus just a trivial setup is
 		 * applied here.
 		 */
-		usleep_range(12000, 12100);
+		//usleep_range(12000, 12100);
+		msleep(50);
 
 	} else {
 		ret = readl_poll_timeout(ctl_addr, tmp, (tmp & pdn_ack) == 0,
